src/wallet/rpcwallet.cpp

```diff
@@ -3768,7 +3769,9 @@ UniValue z_viewtransaction(const UniValue& params, bool fHelp)
         auto op = res->second;
         auto wtxPrev = pwalletMain->mapWallet.at(op.hash);
 
-        auto decrypted = wtxPrev.DecryptSaplingNote(op).get();
+        // We don't need to check the leadbyte here: if wtx exists in
+        // the wallet, it must have already passed the leadbyte check
```

Comment:
This is not necessarily true; the transaction might have been received by the wallet from the mempool before the grace period ended, but not mined until afterwards.

Reply:
if we're sure that all txs in the wallet have to pass `ContextualCheckTransaction()`, then we can be sure that post-Canopy txs in the wallet have leadbyte `0x02`:

```C++
bool ContextualCheckTransaction(...) {
...
// ZIP 212: Check that the note plaintexts use the v2 note plaintext
// version.
// This check compels miners to switch to the new plaintext version
// and overrides the grace period in plaintext_version_is_valid()
if (canopyActive != (encPlaintext->get_leadbyte() == 0x02)) {
    return state.DoS(
        DOS_LEVEL_BLOCK,
        error("CheckTransaction(): coinbase output description has invalid note plaintext version"),
        REJECT_INVALID,
        "bad-cb-output-desc-invalid-note-plaintext-version");
}
...
}
```

Reply:
This check only enforces that logic for shielded outputs in coinbase transactions, so str4d's original scenario is still plausible. However, I think we should feel comfortable overlooking this boundary condition with the grace period.